src/main/java/com/billyyccc/mssqlclient/impl/codec/ExtendedQueryCommandCodec.java

```diff
@@ -0,0 +1,236 @@
+    }
+
+    int packetLen = packet.writerIndex() - packetLenIdx + 2;
+    packet.setShort(packetLenIdx, packetLen);
+
+    chctx.writeAndFlush(packet);
+  }
+
+  private String parseParamDefinitions(Tuple params) {
+    StringBuilder stringBuilder = new StringBuilder();
+    for (int i = 0; i < params.size(); i++) {
+      Object param = params.getValue(i);
+      stringBuilder.append("@P").append(i + 1).append(" ");
+      stringBuilder.append(inferenceParamDefinitionByValueType(param));
+      if (i != params.size() - 1) {
+        stringBuilder.append(",");
+      }
+    }
+    return stringBuilder.toString();
+  }
+
+  private void encodeNVarcharParameter(ByteBuf payload, String value) {
+    payload.writeByte(0x00); // name length
+    payload.writeByte(0x00); // status flags
+    payload.writeByte(MSSQLDataTypeId.NVARCHARTYPE_ID);
+    payload.writeShortLE(8000); // maximal length
+    payload.writeByte(0x09);
+    payload.writeByte(0x04);
+    payload.writeByte(0xd0);
+    payload.writeByte(0x00);
+    payload.writeByte(0x34); // Collation for param definitions TODO always this value?
+    writeUnsignedShortLenVarChar(payload, value);
+  }
+
+  private String inferenceParamDefinitionByValueType(Object value) {
+    if (value instanceof Byte) {
+      return "tinyint";
+    } else if (value instanceof Short) {
+      return "smallint";
+    } else if (value instanceof Integer) {
+      return "int";
+    } else if (value instanceof Long) {
+      return "bigint";
+    } else if (value == null || value instanceof String) {
+      return "nvarchar(4000)";
+    } else if (value instanceof LocalDate) {
+      return "date";
+    } else if (value instanceof LocalTime) {
+      return "time";
+    } else {
+      throw new UnsupportedOperationException("Unsupported type");
+    }
+  }
+
+  private void encodeParamValue(ByteBuf payload, Object value) {
+    if (value == null) {
+      encodeNullParameter(payload);
+    } else if (value instanceof Byte) {
+      encodeIntNParameter(payload, 1, value);
+    } else if (value instanceof Short) {
+      encodeIntNParameter(payload, 2, value);
+    } else if (value instanceof Integer) {
+      encodeIntNParameter(payload, 4, value);
+    } else if (value instanceof Long) {
+      encodeIntNParameter(payload, 8, value);
+    } else if (value instanceof String) {
+      encodeNVarcharParameter(payload, (String) value);
+    } else if (value instanceof LocalDate) {
+      throw new UnsupportedOperationException("Unsupported type");
+    } else if (value instanceof LocalTime) {
+      throw new UnsupportedOperationException("Unsupported type");
+    } else {
+      throw new UnsupportedOperationException("Unsupported type");
+    }
+  }
+
+  private void encodeNullParameter(ByteBuf payload) {
+    payload.writeByte(0x00);
+    payload.writeByte(0x00);
+    payload.writeByte(MSSQLDataTypeId.NVARCHARTYPE_ID);
+    payload.writeShortLE(8000); // maximal length
+    payload.writeByte(0x09);
```

Comment:
you should colaesce these calls into a writeLong for better performance

Reply:
this is a good point, but I think we may revisit this later because some options are presently hardcoded for simplier implementation, I believe we can optimize these logics when we have a better outline of this project.

Reply:
ok fine